### ISSUE TITLE
Crash Info

### ISSUE THREAD

**Issue by yanyiwu** [samuel/go-thrift#34]
```
runtime: g18: leftover defer sp=0x0 pc=0x0
    defer 0xc208140000 sp=0xc20829fce8 pc=0x776a02
    defer 0xc208416040 sp=0x0 pc=0x0
fatal error: traceback has leftover defers

runtime stack:
runtime.throw(0xa0a070, 0x1d)
    /home/yanyiwu/local/go/src/runtime/panic.go:492 +0x98
runtime.gentraceback(0x776dd0, 0xc20829e570, 0x0, 0xc208092120, 0x0, 0x0, 0x7fffffff, 0xa93290, 0x7f19cde98cc0, 0x0, ...)
    /home/yanyiwu/local/go/src/runtime/traceback.go:414 +0x6dc
runtime.copystack(0xc208092120, 0x4000)
    /home/yanyiwu/local/go/src/runtime/stack1.go:553 +0x187
runtime.newstack()
    /home/yanyiwu/local/go/src/runtime/stack1.go:725 +0xaa4
runtime.morestack()
    /home/yanyiwu/local/go/src/runtime/asm_amd64.s:309 +0x7e

goroutine 18 [copystack]:
github.com/samuel/go-thrift/thrift.(*decoder).readValue(0xc2082808c0, 0xc208280b0c, 0x7c58e0, 0xc208280ba0, 0xd6)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/decoder.go:51 fp=0xc20829e578 sp=0xc20829e570
github.com/samuel/go-thrift/thrift.(*decoder).readValue(0xc2082808c0, 0xc208280b0c, 0x7c4f20, 0xc208280b90, 0xd6)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/decoder.go:164 +0x1521 fp=0xc20829ea28 sp=0xc20829e578
github.com/samuel/go-thrift/thrift.(*decoder).readValue(0xc2082808c0, 0xc20846200c, 0x7c5a00, 0xc208462028, 0xd6)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/decoder.go:164 +0x1521 fp=0xc20829eed8 sp=0xc20829ea28
github.com/samuel/go-thrift/thrift.(*decoder).readValue(0xc2082808c0, 0xc20841620f, 0x7c7620, 0xc208416250, 0xd7)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/decoder.go:212 +0x272d fp=0xc20829f388 sp=0xc20829eed8
github.com/samuel/go-thrift/thrift.(*decoder).readValue(0xc2082808c0, 0xc20847a00c, 0x7c59a0, 0xc20847a010, 0xd6)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/decoder.go:164 +0x1521 fp=0xc20829f838 sp=0xc20829f388
github.com/samuel/go-thrift/thrift.(*decoder).readValue(0xc2082808c0, 0xc20847a00c, 0x8bd740, 0xc20847a010, 0xd9)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/decoder.go:164 +0x1521 fp=0xc20829fce8 sp=0xc20829f838
github.com/samuel/go-thrift/thrift.DecodeStruct(0x7f19cf032dc0, 0xc208047680, 0x7c5760, 0xc20847a010, 0x0, 0x0)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/decoder.go:43 +0x460 fp=0xc20829fda8 sp=0xc20829fce8
github.com/samuel/go-thrift/thrift.(*clientCodec).ReadResponseBody(0xc2080a21b0, 0x7c5760, 0xc20847a010, 0x0, 0x0)
    /home/yanyiwu/golang/src/github.com/samuel/go-thrift/thrift/client.go:159 +0xa5 fp=0xc20829fe00 sp=0xc20829fda8
net/rpc.(*Client).input(0xc208094120)
    /home/yanyiwu/local/go/src/net/rpc/client.go:141 +0xa62 fp=0xc20829ffd8 sp=0xc20829fe00
runtime.goexit()
    /home/yanyiwu/local/go/src/runtime/asm_amd64.s:2403 +0x1 fp=0xc20829ffe0 sp=0xc20829ffd8
created by net/rpc.NewClientWithCodec
    /home/yanyiwu/local/go/src/net/rpc/client.go:201 +0xdb
```

Does anyone else run into this bug?

**Comment by samuel** [comment]
Which version of Go are you using? This seems like an issue in the runtime as the Thrift package doesn't use unsafe or cgo anywhere.

There's a few instances of this error in the go issue that have been fixed (for instance this for Go 1.3 https://code.google.com/p/go/issues/detail?id=8132)

**Comment by yanyiwu** [comment]
Thanks, But I run its `x.go` (as following) is ok.
```
package main

import "runtime"

func main() {
	func() {
		defer func() {
			runtime.GC()
			recover()
		}()
		var x [8192]byte
		func(x [8192]byte) {
			defer func() {
				if err := recover(); err != nil {
					panic("wrap: " + err.(string))
				}
			}()
			panic("bad")
		}(x)
	}()
}
```

I use go based on its master branch (commit: a877e81caa23035ec09b4a1b3ad5553617b41dae)

**Comment by samuel** [comment]
Oh, sorry, I didn't mean that it's the exact error from the issue but that the look of the error makes me think it's internal to the Go runtime.

Please try your code with the Go 1.4.2 release to see if it still happens.

**Comment by yanyiwu** [comment]
OK, Thanks ,I will try it.

**Comment by yanyiwu** [comment]
It is fixed when I use Go 1.4.2
